artiq/applets/big_number.py

```diff
@@ -26,54 +30,70 @@ def __init__(self, args, ctl):
+        self.display_widget.setAlignment(
+            QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
+        self.display_widget.doubleClicked.connect(self.start_edit)
+        self.addWidget(self.display_widget)
 
         self.edit_widget = QCancellableLineEdit()
-        self.edit_widget.setValidator(QtGui.QDoubleValidator())
-        self.edit_widget.setAlignment(QtCore.Qt.AlignRight)
-        self.edit_widget.editCancelled.connect(self.cancel_edit)
+        self.edit_widget.setFont(font)
+        self.edit_widget.setAlignment(
+            QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
+        self.edit_widget.editCancelled.connect(
+            lambda: self.setCurrentWidget(self.display_widget))
         self.edit_widget.returnPressed.connect(self.confirm_edit)
         self.addWidget(self.edit_widget)
 
-        font = QtGui.QFont()
-        font.setPointSize(60)
-        self.edit_widget.setFont(font)
+        float_regexp = r"(([+-]?\d+(\.\d*)?|\.\d+)([eE][+-]?\d+)?)"
+        self.regexp = "^" + float_regexp + r"( [a-zA-Z]*)?$"
+        self.edit_widget.setValidator(
+            QtGui.QRegExpValidator(QtCore.QRegExp(self.regexp)))
 
-        self.setCurrentWidget(self.lcd_widget)
+        self.setCurrentWidget(self.display_widget)
 
     def start_edit(self):
-        # QLCDNumber value property contains the value of zero
-        # if the displayed value is not a number.
-        self.edit_widget.setText(str(self.lcd_widget.value()))
+        if self.value is None:
+            self.edit_widget.setText("---")
+        else:
+            unit = self.metadata.get("unit", "")
+            scale = getattr(units, unit, 1)
+            value = self.value/scale
+            self.edit_widget.setText(f"{value} {unit}".strip())
         self.edit_widget.selectAll()
         self.edit_widget.setFocus()
         self.setCurrentWidget(self.edit_widget)
 
     def confirm_edit(self):
-        value = float(self.edit_widget.text())
-        self.ctl.set_dataset(self.dataset_name, value)
-        self.setCurrentWidget(self.lcd_widget)
-
-    def cancel_edit(self):
-        self.setCurrentWidget(self.lcd_widget)
+        entry = re.match(self.regexp, self.edit_widget.text())
+        value = float(entry.group(1))
+        unit = entry.group(5).strip() if entry.group(5) else ""
+        scale = getattr(units, unit, 1)
+        precision = self.metadata.get("precision")
+        self.ctl.set_dataset(self.dataset_name, value*scale, unit=unit,
+                             precision=precision)
+        self.setCurrentWidget(self.display_widget)
 
     def data_changed(self, value, metadata, persist, mods):
+        self.metadata = metadata.get(self.dataset_name, {})
         try:
-            n = float(value[self.dataset_name])
+            self.value = float(value[self.dataset_name])
+            n = short_format(self.value, self.metadata)
         except (KeyError, ValueError, TypeError):
             n = "---"
```

Comment:
"---" cannot be submitted, the widget will not exit editing mode. Same with an empty string. If the user cancels at an invalid state, the original dataset should be kept as is.

Reply:
What do you mean by this? From what I understand, this is referring to editing a value from the dashboard while also edit mode in widget is currently open, any entry will not be save including the valid ones. There is a patch for this that will be issued in a separate PR.

> This happens when user is still inside edit_widget but tried to go somewhere else without cancelling the edit or pressing enter (e.g. tried to change the value of dataset in gui). Proposed solution updates the edit_widget to shift to display when data_changed happens while still inside edit_widget. Will just transfer this patch to a separate PR if we want to retain this.

Reply:
No I am not talking about that specifically. Here are the precise steps to reproduce:

1. Set a dataset to a string value (not a number)
2. Call the big_number applet with this dataset as input
3. Click on the "---" to edit it
4. At this point, pressing enter did not leave edit mode, and when deleting all the '-' chars, pressing enter also did not work. I was forced to override the string value with a number in order to exit the applet.

Reply:
Updated. Pressing enter when input is invalid will revert to previous valid value or "---" on start. Added a warning log as well.